relayer/forwarder.go

```diff
@@ -0,0 +1,123 @@
+package relayer
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+
+	rcommon "github.com/athanorlabs/go-relayer/common"
+	"github.com/athanorlabs/go-relayer/impls/gsnforwarder"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+)
+
+func createForwarderSignature(
+	ctx context.Context,
+	claimerEthKey *ecdsa.PrivateKey,
+	ec *ethclient.Client,
+	relayerFeeWei *big.Int,
+	swapFactoryAddress ethcommon.Address,
+	forwarderAddress ethcommon.Address,
+	swap *contracts.SwapFactorySwap,
+	secret *[32]byte,
+) ([]byte, error) {
+
+	if swap.Claimer != ethcrypto.PubkeyToAddress(claimerEthKey.PublicKey) {
+		return nil, fmt.Errorf("signing key does not match claimer %s", swap.Claimer)
+	}
+
+	forwarder, domainSeparator, err := getForwarderAndDomainSeparator(ctx, ec, forwarderAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce, err := forwarder.GetNonce(&bind.CallOpts{Context: ctx}, swap.Claimer)
+	if err != nil {
+		return nil, err
+	}
+
+	forwarderReq, err := createForwarderRequest(
+		nonce,
+		relayerFeeWei,
+		swapFactoryAddress,
+		swap,
+		secret,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	digest, err := rcommon.GetForwardRequestDigestToSign(forwarderReq, *domainSeparator, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get forward request digest: %w", err)
+	}
+
+	signature, err := rcommon.NewKeyFromPrivateKey(claimerEthKey).Sign(digest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign forward request digest: %w", err)
+	}
+
+	return signature, nil
+}
+
+// createForwarderRequest creates the forwarder request, which we sign the digest of.
+func createForwarderRequest(
+	nonce *big.Int,
+	relayerFeeWei *big.Int,
+	swapFactoryAddress ethcommon.Address,
+	swap *contracts.SwapFactorySwap,
+	secret *[32]byte,
+) (*gsnforwarder.IForwarderForwardRequest, error) {
+
+	calldata, err := getClaimRelayerTxCalldata(relayerFeeWei, swap, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &gsnforwarder.IForwarderForwardRequest{
+		From:           swap.Claimer,
+		To:             swapFactoryAddress,
+		Value:          big.NewInt(0),
+		Gas:            big.NewInt(200000), // TODO: fetch from ethclient?
```

Comment:
can you set this to 52000 actually? a claim shouldn't be more than that; this will prevent the relayer from maybe doing something with our excess gas.

Reply:
a claim is usually just under 40000 actually, unless it's the very first claim in a contract

Reply:
Constant in place